grant-project/zcash-grant-system
# Better Milestone Validation

Closes #192, as well as some untracked issues.

### What This Does

Removes some of the confusing "smart" behavior of milestones and validation / payout percentages. Percentage payout amount field no longer forces itself into a valid value, or divides the milestone before it when adding one, and the errors are never silent. This has the downside of showing an error immediately upfront, but has the upside of the user's proposal never being in a silent, invalid state. Payout also no longer forces numeric input, instead opting to show an error when the input is invalid.

It's worth noting that the milestone percentage field is a string in the db, and does zero validation of this field. This didn't cause any issues in testing, but we may want to adjust this behavior since it's a little weird.

### Steps to Test

* Create a proposal, go to the review step with a milestone missing fields. Confirm the error shows correctly.
* Create a milestone and set an invalid number for payout percent. Confirm an error shows correctly.
* Create a set of milestones where they add up to a non-100 number. Confirm an error shows correctly.
* Create a valid set of milestones, confirm you can submit the proposal.
